app/src/main/java/info/plateaukao/einkbro/epub/EpubManager.kt

```diff
@@ -57,17 +57,23 @@ class EpubManager(private val context: Context) : KoinComponent {
             if (bookName != null && chapterName != null) {
                 val progressDialog = ProgressDialog(activity, R.style.TouchAreaDialog).apply {
                     setTitle(R.string.saving_epub)
+                    setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
+                    isIndeterminate = false
+                    max = 100
                     show()
                 }
 
                 val rawHtml = ninjaWebView.getRawReaderHtml()
+                progressDialog.progress = 5
+
                 internalSaveEpub(
                     isNewFile,
                     fileUri,
                     rawHtml,
                     bookName,
                     chapterName,
                     ninjaWebView.url ?: "",
+                    progressDialog,
```

Comment:
it's not a good way to send progressDialog as argument into `internalSaveEpub`, and `saveImageReources()`. These two functions should not be aware of any UI component, or try to do any UI updates.
Instead, they could accept a callback funtion to emit the progress percentage, and let the top layer decide what to behave on UI part.
I'll have an implementation based on your pull request. Thank you for sending the pull request.

Reply:
Sounds good, please keep in mind that we might want to pull the images in a few background threads to have them work on their (possible) timeouts in parallel.

I think it might be best to:
- do a first pass over the HMTL to count the number of images, then we know the N of downloads to attempt
- do a second pass, where a callback can +1 the number of completed downloads, and while here, where we have access to the mimetype, add the resource with the proper extension to the epub directly.

That's in any case the patch I'm going to work on next once this has landed. Would you prefer the ebook to get passed down to processHMTL and getImage, or should this be callback based? (wouldn't that incur a lot of copying around of images, which could be many MB big?)